File: apps/desktop/src/shared/formats.ts
/**
 * The creative-file format registry — one source of truth for every format
 * Chronicle can capture, and for how each one is previewed and displayed.
 *
 * Imported by main, preload, and the renderer, so this module stays free of
 * Node built-ins and of any byte-level parsing: it is pure data plus string
 * helpers. Format-specific *behavior* (metadata readers, preview generators)
 * lives in `src/main/formats/`, keyed by the `id` values declared here.
 *
 * Adding a format:
 *   1. add one entry below,
 *   2. register its metadata reader / preview generator in `src/main/formats/`,
 *   3. add its prompt section + adapter in the AI service when annotation is
 *      implemented (until then `aiFormat: null` keeps its jobs queued).
 * Nothing else in the app should ever branch on a file extension.
 */

export type FormatId = 'png' | 'jpg' | 'svg' | 'psd' | 'psb' | 'obj' | 'step' | 'blend'

/** How the renderer obtains something to show for a version. */
export type PreviewStrategy =
  /** Chromium decodes the stored bytes directly (PNG/JPG/SVG). */
  | 'native'
  /** Main derives a displayable image from the stored bytes and caches it. */
  | 'derived'
  /** No still image is available; lists show the format placeholder. */
  | 'none'

/** Which renderer component shows a version's visual. */
export type ViewerKind =
  /** A bitmap in an <img> element. */
  | 'raster'
  /** Vector markup in an <img> element (never inlined — see media.ts). */
  | 'svg'
  /** Interactive 3D mesh view. */
  | 'mesh3d'

export interface FormatDescriptor {
  id: FormatId
  /** User-facing name, used by the project file-type toggles. */
  label: string
  /** Lowercase, dot-prefixed extensions. The first one is the canonical form. */
  extensions: readonly string[]
  /** Content type used when serving the original bytes to the renderer. */
  mediaType: string
  preview: PreviewStrategy
  viewer: ViewerKind
  /**
   * Icon name (see renderer `Icon.tsx`) shown when no preview image exists.
   * Declared here so the renderer never maps file types itself.
   */
  icon: 'image' | 'layers' | 'cube' | 'architecture' | 'shapes'
  /**
   * How a still preview fills its frame. Photographic content is cropped to
   * fill ('cover'); vector artwork and rendered geometry must stay whole
   * ('contain'), because cropping a diagram or a model loses its subject.
   */
  fit: 'cover' | 'contain'
  /**
   * The C3 `format` value to send with an annotation request, or null for a
   * format Chronicle captures but never asks the AI service to annotate. Null
   * keeps captured versions' annotation jobs queued instead of failing them
   * (spec F4). Every format declared here has an adapter as of POST-02; whether
   * the *running* service actually accepts it is answered by `GET /capabilities`
   * (see `src/main/ai/capabilities.ts`), never assumed from this field.
   */
  aiFormat: string | null
  /** Per-format capture ceiling; falls back to the global cap when absent. */
  maxBytes?: number
}

/**
 * Declared in the roadmap order from docs/challenge/RESEARCH.md. PNG and JPG
 * are the MVP baseline; POST-02 added the rest, each annotated through its own
 * safe local-extraction adapter in the AI service rather than by sending
 * opaque project bytes to a provider.
 */
export const FORMATS: readonly FormatDescriptor[] = [
  {
    id: 'png',
    label: 'PNG',
    extensions: ['.png'],
    mediaType: 'image/png',
    preview: 'native',
    viewer: 'raster',
    icon: 'image',
    fit: 'cover',
    aiFormat: 'png',
  },
  {
    id: 'jpg',
    label: 'JPG / JPEG',
    extensions: ['.jpg', '.jpeg'],
    mediaType: 'image/jpeg',
    preview: 'native',
    viewer: 'raster',
    icon: 'image',
    fit: 'cover',
    aiFormat: 'jpg',
  },
  {
    id: 'svg',
    label: 'SVG',
    extensions: ['.svg'],
    mediaType: 'image/svg+xml',
    preview: 'native',
    viewer: 'svg',
    icon: 'shapes',
    fit: 'contain',
    aiFormat: 'svg',
  },
  {
    id: 'psd',
    label: 'Photoshop PSD',
    extensions: ['.psd'],
    mediaType: 'image/vnd.adobe.photoshop',
    preview: 'derived',
    viewer: 'raster',
    icon: 'layers',
    fit: 'cover',
    aiFormat: 'psd',
  },
  {
    id: 'psb',
    label: 'Photoshop PSB',
    extensions: ['.psb'],
    mediaType: 'image/vnd.adobe.photoshop',
    preview: 'derived',
    viewer: 'raster',
    icon: 'layers',
    fit: 'cover',
    aiFormat: 'psb',
  },
  {
    id: 'obj',
    label: 'OBJ 3D model',
    extensions: ['.obj'],
    mediaType: 'model/obj',
    preview: 'derived',
    viewer: 'mesh3d',
    icon: 'cube',
    fit: 'contain',
    aiFormat: 'obj',
  },
  {
    // STEP needs a CAD kernel to tessellate, which runs in the renderer's
    // 3D viewer rather than in main, so there is no still thumbnail.
    id: 'step',
    label: 'STEP CAD',
    extensions: ['.step', '.stp'],
    mediaType: 'model/step',
    preview: 'none',
    viewer: 'mesh3d',
    icon: 'architecture',
    fit: 'contain',
    aiFormat: 'step',
  },
  {
    id: 'blend',
    label: 'Blender BLEND',
    extensions: ['.blend'],
    mediaType: 'application/x-blender',
    preview: 'derived',
    viewer: 'raster',
    icon: 'cube',
    fit: 'cover',
    aiFormat: 'blend',
  },
]

const BY_ID = new Map(FORMATS.map((format) => [format.id, format]))
const BY_EXTENSION = new Map(
  FORMATS.flatMap((format) => format.extensions.map((extension) => [extension, format] as const)),
)

/** Every capturable extension, lowercase and dot-prefixed (C4 source data). */
export const SUPPORTED_EXTENSIONS: readonly string[] = [...BY_EXTENSION.keys()]

/** Lowercase, dot-prefixed extension of a path, or '' when it has none. */
export function extensionOf(filePath: string): string {
  const base = filePath.slice(Math.max(filePath.lastIndexOf('/'), filePath.lastIndexOf('\\')) + 1)
  const dot = base.lastIndexOf('.')
  return dot <= 0 ? '' : base.slice(dot).toLowerCase()
}

/** The descriptor for a path's extension, or null when it is not supported. */
export function formatForPath(filePath: string): FormatDescriptor | null {
  return BY_EXTENSION.get(extensionOf(filePath)) ?? null
}

/** The descriptor for a known id. Throws for an unknown id (a programming error). */
export function formatById(id: FormatId): FormatDescriptor {
  const format = BY_ID.get(id)
  if (!format) throw new Error(`Unknown format id: ${id}`)
  return format
}

/** Narrowing guard for values arriving from IPC, URLs, or persisted rows. */
export function isFormatId(value: unknown): value is FormatId {
  return typeof value === 'string' && BY_ID.has(value as FormatId)
}

/**
 * True when the AI service accepts this format today. False means capture,
 * preview, timeline, restore, and keyword search all work while the version's
 * annotation job waits in the queue for POST-02 support.
 */
export function supportsAnnotation(format: FormatDescriptor): boolean {
  return format.aiFormat !== null
}

File: services/ai/chronicle_ai/formats.py
"""Per-format annotation adapters.

The C3 ``format`` field selects an adapter here; nothing else in the service
branches on a file type. An adapter declares the media type its inputs carry,
which prompt wording to use, and — optionally — a local preparation step that
turns opaque bytes into provider-safe evidence (structure plus at most a derived
preview image), so proprietary container bytes never reach the provider.

Adding a format: add one ``FormatAdapter`` below, add its prompt sections to
``packages/prompts/version-annotation.md``, and widen the ``SupportedFormat``
literal in ``schemas.py``. The desktop app keeps that format's annotation jobs
queued until this registry accepts it (see /capabilities).
"""

from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from typing import Callable, Protocol, Sequence


class PreparedAnnotation(Protocol):
    """Local evidence an adapter produces for the provider call."""

    @property
    def context(self) -> str: ...

    @property
    def images(self) -> Sequence[object]: ...

    @property
    def confidence_limit(self) -> float | None: ...


@dataclass(frozen=True)
class FormatAdapter:
    """Everything the annotation pipeline needs to know about one format."""

    format: str
    #: Media type the C3 input must carry for this format.
    media_type: str
    #: Prompt operation prefix in packages/prompts/version-annotation.md.
    prompt_operation: str
    #: Local extraction step, or None to send the original bytes as an image.
    prepare: Callable[..., PreparedAnnotation] | None = None


def _image_adapter(format_name: str, media_type: str) -> FormatAdapter:
    """Formats a vision model reads directly: no local extraction needed."""

    return FormatAdapter(format=format_name, media_type=media_type, prompt_operation="")


@lru_cache(maxsize=1)
def _registry() -> dict[str, FormatAdapter]:
    # Adapter modules are imported on first use, not at module import time:
    # they depend on the request schemas, which in turn read this registry.
    from .future_formats import (
        prepare_blend_annotation,
        prepare_obj_annotation,
        prepare_step_annotation,
        prepare_svg_annotation,
    )
    from .psd_adapter import prepare_psd_annotation
    from .psd_adapter import prepare_psb_annotation

    return {
        "png": _image_adapter("png", "image/png"),
        "jpg": _image_adapter("jpg", "image/jpeg"),
        "jpeg": _image_adapter("jpeg", "image/jpeg"),
        "psd": FormatAdapter(
            format="psd",
            media_type="image/vnd.adobe.photoshop",
            prompt_operation="PSD ",
            prepare=prepare_psd_annotation,
        ),
        "psb": FormatAdapter(
            format="psb",
            media_type="image/vnd.adobe.photoshop",
            prompt_operation="PSB ",
            prepare=prepare_psb_annotation,
        ),
        "svg": FormatAdapter(
            format="svg",
            media_type="image/svg+xml",
            prompt_operation="SVG ",
            prepare=prepare_svg_annotation,
        ),
        "blend": FormatAdapter(
            format="blend",
            media_type="application/x-blender",
            prompt_operation="BLEND ",
            prepare=prepare_blend_annotation,
        ),
        "obj": FormatAdapter(
            format="obj",
            media_type="model/obj",
            prompt_operation="OBJ ",
            prepare=prepare_obj_annotation,
        ),
        "step": FormatAdapter(
            format="step",
            media_type="model/step",
            prompt_operation="STEP ",
            prepare=prepare_step_annotation,
        ),
    }


def adapter_for(format_name: str) -> FormatAdapter | None:
    """The adapter for a format, or None when the service cannot annotate it."""

    return _registry().get(format_name)


def supported_formats() -> tuple[str, ...]:
    """Formats this build can annotate, in declaration order (/capabilities)."""

    return tuple(_registry())


def media_types() -> tuple[str, ...]:
    """Media types accepted across every registered format."""

    return tuple(dict.fromkeys(adapter.media_type for adapter in _registry().values()))
